GVFS/GVFS.Common/Tracing/TelemetryDaemonEventListener.cs

```diff
@@ -0,0 +1,163 @@
+
+        internal static string CreateJsonMessage(
+            string vfsVersion,
+            string providerName,
+            string enlistmentId,
+            string mountId,
+            string eventName,
+            Guid activityId,
+            Guid parentActivityId,
+            EventLevel level,
+            Keywords keywords,
+            EventOpcode opcode,
+            string payload)
+        {
+            var message = new Dictionary<string, object>
+            {
+                ["version"] = vfsVersion,
+                ["providerName"] = providerName,
+                ["eventName"] = eventName,
+                ["eventLevel"] = ((int)level).ToString(),
+                ["eventOpcode"] = ((int)opcode).ToString(),
+                ["payload"] = new Dictionary<string, string>
+                {
+                    ["enlistmentId"] = enlistmentId,
+                    ["mountId"] = mountId,
+                    ["json"] = payload,
+                },
+
+                // TODO: do we need these?
+                // ETW-only properties
+                ["etw.activityId"] = activityId.ToString("D"),
+                ["etw.parentActivityId"] = parentActivityId.ToString("D"),
+            };
+
+            return JsonConvert.SerializeObject(message);
+        }
+
+        protected override void RecordMessageInternal(
+            string eventName,
+            Guid activityId,
+            Guid parentActivityId,
+            EventLevel level,
+            Keywords keywords,
+            EventOpcode opcode,
+            string payload)
+        {
+            string message = CreateJsonMessage(
+                this.vfsVersion,
+                this.providerName,
+                this.enlistmentId,
+                this.mountId,
+                eventName,
+                activityId,
+                parentActivityId,
+                level,
+                keywords,
+                opcode,
+                payload);
+
+            this.SendMessage(message);
+        }
+
+        private static string GetConfigValue(string gitBinRoot, string configKey)
+        {
+            string value = string.Empty;
+            string error;
+
+            GitProcess.ConfigResult result = GitProcess.GetFromSystemConfig(gitBinRoot, configKey);
+            if (!result.TryParseAsString(out value, out error, defaultValue: string.Empty) || string.IsNullOrWhiteSpace(value))
+            {
+                result = GitProcess.GetFromGlobalConfig(gitBinRoot, configKey);
+                result.TryParseAsString(out value, out error, defaultValue: string.Empty);
+            }
+
+            return value.TrimEnd('\r', '\n');
+        }
+
+        private void SendMessage(string message)
+        {
+            // Create pipe if this is the first message, or if the last connection broke for any reason
+            if (this.pipeClient == null)
+            {
+                string pipeName = GVFSPlatform.Instance.GetTelemetryNamedPipeName();

```

Comment:
Okay now I see why you added this method to the platform interface.  What are you thoughts on having the method in the `GVFSPlatform` be 
```
void AddTelemetryListenersIfEnabled(List<EventListener> listeners, string providerName, string enlistmentId, string mountId)
```
and the platform can add what it wants.  For windows it would add both for now, mac nothing now but could only be the daemon listener.

Then the `CreateIfEnabled` could take the `pipeName` that the platform passes to it and store it for use here?

I hope that makes sense.  Not sure how much cleaner this would be.

Reply:
Yes that makes sense. I've tried your idea and it looks nicer IMO :+1:

I've replaced the `CreatePlatformTelemetryListener[IfEnabled]` method with one that can return possibly multiple `EventListener`s. These listeners (zero or many) are attached to the `JsonTracer` if any are returned.

```csharp
// GVFSPlatform.cs
IEnumerable<EventListener> CreateTelemetryListeners(...);

// JsonTracer.cs
IEnumerable<EventListener> listeners = GVFSPlatform.Instance.CreateTelemetryListeners(...);
this.listeners.AddRange(listeners);
```

Windows returns both (if enabled) and Mac will eventually return one (the telemetry daemon one).

Note: I went with a `Create___()` method and letting the `JsonTracer` add the returned listeners itself, rather than have the tracer have to hand over it's internal `List<>` and couple the platform code to the tracer's internal data structure.

Reply:
cool.

Reply:
Looks good thanks!